components/fxa-client/ios/FxAClient/FxAccountDeviceConstellation.swift

```diff
@@ -0,0 +1,192 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+public protocol DeviceConstellationProtocol {
+    // Get local + remote devices.
+    func state() -> ConstellationState?
+    // Refresh the list of remote devices.
+    // Dispatchs a notification to NSNotificationCenter, using the name `.constellationStateUpdate`.
+    // The updated `ConstellationState` can be found in `userInfo` under the key `newState`.
+    func refreshState()
+
+    func setLocalDeviceName(name: String)
+
+    // Poll for events we might have missed (e.g. no push notification)
+    func pollForEvents(completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void)
+    // Send an event to another device such as Send Tab.
+    func sendEventToDevice(targetDeviceId: String, e: DeviceEventOutgoing)
+
+    // Register our push subscription with the FxA server.
+    func setDevicePushSubscription(sub: DevicePushSubscription)
+    // Used by Push when receiving a message.
+    func processRawIncomingDeviceEvent(
+        pushPayload: String,
+        completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void
+    )
+}
+
+public extension Notification.Name {
+    static let constellationStateUpdate = Notification.Name("constellationStateUpdate")
+}
+
+public struct ConstellationState {
+    public let localDevice: Device?
+    public let remoteDevices: [Device]
+}
+
+public class DeviceConstellation: DeviceConstellationProtocol {
+    var constellationState: ConstellationState?
+    let account: FirefoxAccount
+
+    init(account: FirefoxAccount) {
+        self.account = account
+    }
+
+    public func state() -> ConstellationState? {
+        return constellationState
+    }
+
+    public func refreshState() {
+        DispatchQueue.global().async {
+            FxALog.info("Refreshing device list...")
+            do {
+                let devices = try self.handleAccountAuthErrors { acct in
+                    try acct.fetchDevicesSync()
+                }
+                let localDevice = devices.first { $0.isCurrentDevice }
+                if localDevice?.subscriptionExpired ?? false {
```

Comment:
Note to @garvankeeley: you should probably interrogate this value from time to time on iOS.